casper/src/main/resources/PoS.rho

```diff
@@ -1,180 +1,165 @@
-                      | for (@(true, vault) <- vaultCh) {
-                        @RevVault!("deployerAuthKey", *authKeyCh)
-                        | for (userAuthKey <- authKeyCh){
-                          @vault!("transfer", posRevAddress, amount, *userAuthKey, *return)
-                        }
-                      }
-                    }
-                  }
-                }
-              }
+            contract PoS (@"pay", @amount, returnCh) = {
+              new vaultCh, transferCh, userCh in {
+                getUser!(*userCh) |
 
-              | paymentStore!([Nil, Nil, Nil, Nil])
-
-              //Make it easy to query the last payment so that validators
-              //can check how much phlo there is for the deploy.
-              | contract PoS(@"lastPayment", return) = {
-                for(@payment <- paymentStore) {
-                  paymentStore!(payment) | return!(payment)
+                for(@userPk <- userCh) {
+                  deposit!(userPk, amount, *returnCh)
                 }
               }
-
-              | contract PoS (@"pay", @amount, return) = {
-                new parametersCh, depositRet in {
-                  deposit!(amount, *depositRet)
-                  | getParameters!(*parametersCh)
-                  | for (@depositResult <- depositRet; @codeHash, @phloRate, @userPk, @timestamp <- parametersCh; lastPayment <- paymentStore) {
-                    match depositResult {
-                      (true, Nil) => { paymentStore!([codeHash, userPk, timestamp, phloRate * amount]) | return!(phloRate *amount) }
-                      (false, err) => { paymentStore!([codeHash, userPk, timestamp, 0]) | return!(0) }
-                    }
-                  }
+            }
+          }
+        } |
+
+        contract deposit (@fromUserPk, @amount, returnCh) = {
+          new vaultCh,
+              revAddressCh,
+              authKeyCh in {
+                stdout!("get deployerauthkey") |
+            @RevVault!("deployerAuthKey", *authKeyCh) |
+            for (@authKey <- authKeyCh) {
+              stdout!({"deployerauthkey" : authKey}) |
+              revAddressOps!("fromPublicKey", fromUserPk, *revAddressCh) |
+              for (@fromRevAddress <- revAddressCh) {
+                stdout!({"address": fromRevAddress}) |
+                @RevVault!("findOrCreate", fromRevAddress, *vaultCh) |
+                for (@(true, fromVault) <- vaultCh) {
+                  stdout!({"vault":fromVault}) |
+                  @fromVault!("transfer", posRevAddress, amount, authKey, *returnCh)
                 }
               }
             }
           }
         }
+      } |
+
+      contract getUser (returnCh) = {
+        new parametersCh in {
+          getParameters!(*parametersCh) |
+          for (@codeHash, @phloRate, @userPk, @timestamp <- parametersCh) {
+            returnCh!(userPk)
+          }
+        }
+      } |
+
+      contract getCurrentUserAddress(returnCh) = {
```

Comment:
Do you use this? It looks like you intended to on line 39, but your TODO doesn't explain why you didn't.

Reply:
it's also used in getCurrentUserVault which is used to create the vault when the PoS instance is created